keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1046 @@
+
+**Support private keys shared across multiple pods**: Sharing private keys and certificates across multiple pods may be required in certain applications.  However, this use case is well-served by Secret projected volume sources.
+
+**Support human-in-the-loop approval for certificate requests**: This KEP assumes that CSR approval and issuance is fully-automated, because this process blocks pod startup for pods that use PodCertificate projected volume sourcs.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+#### Story 1
+
+I'm an application developer building a Kubernetes controller that I want to deploy into my cluster.  I want to make sure that my controller securely authenticates to kube-apiserver using mTLS.
+
+I add a new projected volume to my pod, containing the information that client-go expects at a well-known path.  I set `noAutomountServiceAccountToken` because I want to ensure that my pod cannot use bearer tokens for authentication.
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: default
+  name: pod-certificates-example
+spec:
+  restartPolicy: OnFailure
+  automountServiceAccountToken: false
+  containers:
+  - name: main
+    image: debian
+    command: ['sleep', 'infinity']
+    volumeMounts:
+    - name: kube-apiserver-client-certificate
+      mountPath: /var/run/secrets/kubernetes.io/api-mtls-access
+  volumes:
+  - name: kube-apiserver-client-certificate
+    projected:
+      sources:
+      - podCertificate:
+          signerName: "kubernetes.io/kube-apiserver-client-pod"
+          credentialBundlePath: client-credentials.pem
+      - configMap:
+          localObjectReference: kube-root-ca.crt
+          items:
+          - key: ca.crt
+            path: kube-apiserver-root-certificate.pem
+      - downwardAPI:
+          items:
+          - path: namespace
+            fieldRef:
+              apiVersion: v1
+              fieldPath: metadata.namespace
+```
+
+I ensure that my pod uses client-go's InClusterConfig, and client-go automatically uses the mTLS client certificate to authenticate to kube-apiserver.
+
+### Risks and Mitigations
+
+**Scalability**: In the current design and draft implementation, Kubelet holds
+the private keys for each pod on the node in memory.  If Kubelet restarts, then
```

Comment:
This is just the way the kubelet is structured today.  All data that is written out into a projected volume is held in memory, and the kubelet writes out the data periodically using the atomicwriter utility package.

I think we are probably not going to be able to banish the keys from kubelet memory without going full TPM.  Without a TPM, the kubelet has to hold the key in memory at least temporarily to do the crypto of making the CSR, and during that time period the kubelet's memory might get swapped to disk, etc.

The primary ways we could mitigate the "kubelet restarts -> reissue all certificates" problem are:
1.  Store the keys in a host-level database, probably something based on boltdb.  This *feels* gross, because it seems like it makes the keys easily accessible for any privileged pods or to someone who has root on the node.  However, if you actually think through the analysis, I don't believe it's actually more insecure than allowing the keys to live in Kubelet memory --- root and privileged pods will be able to rummage through Kubelet's memory if they want to.  It is also fairly simple to retrofit into the kubelet's existing projected volume architecture.  This is the alternative we are moving towards in gke-spiffe, which is a CSI driver implementation of this capability.
2. Read the keys back from the pod filesystem.  This is much harder to fit into the existing architecture, and not actually more secure.  This is the alternative we started with in gke-spiffe.

Reply:
The kubelet could defer to something like SoftHSM (or, generically, PKCS#11). And then the PKCS#11 implementation either uses hardware or `mlockall()` to protect the key material from being written to the pagefile.

Reply:
After some discussion in SIG Auth, we decided to defer any sort of HSM capability to a later enhancement.